cargo-binstall fallbacks to cargo-install with a fallback target instead of native target

See [this log](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292).

```
Run taiki-e/install-action@v2
  with:
    tool: bpf-linker
    checksum: true
  env:
    CARGO_TERM_COLOR: always
    CARGO_INCREMENTAL: 0
    CACHE_ON_FAILURE: false
Run bash --noprofile --norc "${GITHUB_ACTION_PATH:?}/main.sh"
  bash --noprofile --norc "${GITHUB_ACTION_PATH:?}/main.sh"
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    CARGO_TERM_COLOR: always
    CARGO_INCREMENTAL: 0
    CACHE_ON_FAILURE: false
    INPUT_TOOL: bpf-linker
    INPUT_CHECKSUM: true
info: host platform: x86_64_linux
info: install-action does not support bpf-linker; fallback to cargo-binstall
info: installing cargo-binstall@latest (1.3.1)
info: downloading https://github.com/cargo-bins/cargo-binstall/releases/download/v1.3.1/cargo-binstall-x86_64-unknown-linux-musl.tgz
info: verifying sha[2](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:2)56 checksum for cargo-binstall-x86_64-unknown-linux-musl.tgz
info: cargo-binstall installed at /home/runner/.cargo/bin/cargo-binstall
+ cargo-binstall binstall -V
1.[3](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:3).1
 WARN Failed to read git credential file /home/runner/.git-credentials err=Os { code: 2, kind: NotFound, message: "No such file or directory" }
 WARN Failed to read git credential file /home/runner/.config/git/credentials err=Os { code: 2, kind: NotFound, message: "No such file or directory" }
 WARN Failed to retrieve token from `gh auth token` err=Custom { kind: Other, error: "process exited with `exit status: 1`" }
 INFO resolve: Resolving package: 'bpf-linker'
 WARN The package bpf-linker v0.9.10 will be installed from source (with cargo)
    Updating crates.io index
 Downloading crates ...
  Downloaded bpf-linker v0.9.10
  Installing bpf-linker v0.9.10
    Updating crates.io index
    Updating crates.io index
 Downloading crates ...
  Downloaded anstyle-parse v0.2.1
  Downloaded clap v[4](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:4).4.4
  Downloaded memchr v2.6.3
  Downloaded ar v0.9.0
  Downloaded camino v1.1.6
  Downloaded aho-corasick v1.0.4
  Downloaded serde_derive v1.0.183
  Downloaded libc v0.2.148
  Downloaded syn v2.0.28
  Downloaded regex-syntax v0.7.[5](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:5)
  Downloaded regex-automata v0.3.8
  Downloaded regex v1.9.5
  Downloaded serde_json v1.0.104
  Downloaded utf8parse v0.2.1
  Downloaded unicode-ident v1.0.11
  Downloaded time-macros v0.2.11
  Downloaded time-core v0.1.1
  Downloaded time v0.3.25
  Downloaded simplelog v0.12.1
  Downloaded serde v1.0.183
  Downloaded semver v1.0.18
  Downloaded ryu v1.0.15
  Downloaded proc-macro2 v1.0.[6](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:6)6
  Downloaded prettyplease v0.2.12
  Downloaded once_cell v1.18.0
  Downloaded log v0.4.20
  Downloaded llvm-sys v1[7](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:7)0.0.0
  Downloaded libloading v0.[8](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:8).0
  Downloaded anyhow v1.0.74
  Downloaded anstream v0.5.0
  Downloaded cargo-platform v0.1.3
  Downloaded anstyle v1.0.1
  Downloaded lazy_static v1.4.0
  Downloaded itoa v1.0.[9](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:10)
  Downloaded cfg-if v1.0.0
  Downloaded anstyle-query v1.0.0
  Downloaded thiserror-impl v1.0.48
  Downloaded thiserror v1.0.48
  Downloaded termcolor v1.1.3
  Downloaded strsim v0.[10](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:11).0
  Downloaded quote v1.0.32
  Downloaded num_threads v0.1.6
  Downloaded llvm-sys v160.1.3
  Downloaded heck v0.4.1
  Downloaded deranged v0.3.7
  Downloaded clap_builder v4.4.4
  Downloaded cc v1.0.82
  Downloaded colorchoice v1.0.0
  Downloaded clap_lex v0.5.0
  Downloaded clap_derive v4.4.2
  Downloaded cargo_metadata v0.18.0
  Downloaded aya-rustc-llvm-proxy v0.9.0
   Compiling serde v1.0.183
   Compiling libc v0.2.148
   Compiling serde_derive v1.0.183
   Compiling proc-macro2 v1.0.66
   Compiling unicode-ident v1.0.[11](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:12)
   Compiling semver v1.0.18
   Compiling memchr v2.6.3
   Compiling quote v1.0.32
   Compiling aho-corasick v1.0.4
   Compiling syn v2.0.28
   Compiling regex-syntax v0.7.5
   Compiling cc v1.0.82
   Compiling regex-automata v0.3.8
   Compiling lazy_static v1.4.0
   Compiling anyhow v1.0.74
   Compiling thiserror v1.0.48
   Compiling thiserror-impl v1.0.48
   Compiling regex v1.9.5
   Compiling camino v1.1.6
   Compiling serde_json v1.0.104
error[E0463]: can't find crate for `core`
  |
  = note: the `x86_64-unknown-linux-musl` target may not be installed
  = help: consider downloading the target with `rustup target add x86_64-unknown-linux-musl`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`

error[E0463]: can't find crate for `compiler_builtins`

error[E0432]: unresolved import `ffi::c_void`
    --> /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/libc-0.2.[14](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:15)8/src/unix/mod.rs:[16](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:17)04:[17](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:18)
     |
1604 |         pub use ::ffi::c_void;
     |                 ^^^^^^^^^^^^^

Some errors have detailed explanations: E0432, E0463.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `libc` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to compile `bpf-linker v0.9.10`, intermediate artifacts can be found at `/tmp/cargo-installP3dSHA`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
ERROR Cargo errored! ExitStatus(unix_wait_status([25](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:27)8[56](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:58)))
ERROR Fatal error:
binstall::subprocess

  × subprocess /home/runner/.rustup/toolchains/nightly-x86_[64](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:66)-unknown-linux-
  │ gnu/bin/cargo install bpf-linker --version 0.9.10 --target x86_64-unknown-
  │ linux-musl --force --locked errored with exit status: 101

Error: Process completed with exit code [70](https://github.com/aya-rs/book/actions/runs/6243912879/job/16950047292#step:6:72).
```

Is the host detection wrong somewhere?

_Originally posted by @tamird in https://github.com/taiki-e/install-action/issues/229_

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -163,6 +163,8 @@ jobs:
     steps:
     - uses: actions/checkout@v4
     - uses: Swatinem/rust-cache@v2
+      with:
+        key: ${{ matrix.os }}
     - name: Build detect-targets
       run: cargo build --bin detect-targets
     - name: Run test in ubuntu

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -153,7 +153,13 @@ jobs:
           alpine test.sh "$TARGET"
 
   detect-targets-ubuntu-test:
-    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-20.04
+          - ubuntu-latest
+    runs-on: ${{ matrix.os }}
     steps:
     - uses: actions/checkout@v4
     - uses: Swatinem/rust-cache@v2

--- a/crates/detect-targets/src/detect/linux.rs
+++ b/crates/detect-targets/src/detect/linux.rs
@@ -108,6 +108,19 @@ You are not meant to run this directly.
         } else {
             None
         }
+    } else if status.code() == Some(127) {
+        // On Ubuntu 20.04 (glibc 2.31), the `--version` flag is not supported
+        // and it will exit with status 127.
+        let status = Command::new(cmd)
+            .arg("/bin/true")
+            .stdin(Stdio::null())
+            .stdout(Stdio::null())
+            .stderr(Stdio::null())
+            .status()
+            .await
+            .ok()?;
+
+        status.success().then_some(Libc::Gnu)
     } else {
         None
     }

